### DISCUSSION TITLE
arbitrary NESTED type conversion

### DISCUSSION THREAD

**Discussion by tesla-cat** [nlohmann/json#3947] [Q&A]
- I saw this example [arbitrary-types-conversions](https://github.com/nlohmann/json#arbitrary-types-conversions)
- but how do you convert json into a nested structure like below?

```cpp
struct Child{
    string name;
};

struct Parent{
    map<string, Child> children;
};

string text = "";

json js = json::parse(text);

Parent p = js.get<Parent>();
```

- (it would even be better to change "struct" to "class")

**Comment by nlohmann** [comment]
Define the conversion for Child. The map of parent is converted automatically.

**Comment by tesla-cat** [comment]
@nlohmann  thanks!  can you please give me some code examples ?

**Reply by nlohmann** [reply]
You will find them on the page you cited.

**Reply by tesla-cat** [reply]
that is only a non-nested structure

my main problem is how do you define `to_json` for both `parent` and `child` ? and what if the `child` has `grand children` ?

**Reply by tesla-cat** [reply]
also, the following code seems pretty redundant

```cpp
void from_json(const json& j, person& p) {
        j.at("name").get_to(p.name);
        j.at("address").get_to(p.address);
        j.at("age").get_to(p.age);
    }
```

why cant we make this automatic ?

imagine you have 100 attibutes, it would be like 

```cpp
void from_json(const json& j, person& p) {
       j.at("attr_1").get_to(p.attr_1);
       j.at("attr_2").get_to(p.attr_2);
       j.at("attr_3").get_to(p.attr_3);
       ......
       j.at("attr_98").get_to(p.attr_98);
       j.at("attr_99").get_to(p.attr_99);
       j.at("attr_100").get_to(p.attr_100);
 }
```

**Reply by nlohmann** [reply]
- You define it for every type.
- This makes it easier: https://json.nlohmann.me/features/arbitrary_types/#simplify-your-life-with-macros

**Reply by tesla-cat** [reply]
thanks, it is helpful, but even with these macros I feel there is still some redundancy, is it not possible to automatically parse json into some structure like in Javascript?

**Reply by gregmarr** [reply]
>thanks, it is helpful, but even with these macros I feel there is still some redundancy, is it not possible to automatically parse json into some structure like in Javascript?

No, not until C++ has reflection, which the committee is working on.
